drivers/include/periph/gpio_ll.h

```diff
@@ -0,0 +1,478 @@
+ *
+ * @note    If the configuration of the Schmitt trigger, the drive strength, or
+ *          the pull resistor strength are not supported, the closest supported
+ *          value will be chosen instead and `0` is returned. The value of these
+ *          fields in @p conf will be updated to the actually chosen value.
+ */
+int gpio_ll_init(gpio_port_t port, uint8_t pin, gpio_conf_t *conf);
+
+/**
+ * @brief   Get the current value of all GPIO pins of the given port as bitmask
+ *
+ * @param[in] port      port to read
+ *
+ * @return              state of all pins as bitmask
+ *
+ * @note    The value of unconfigured pins and pins configured as @ref
+ *          GPIO_DISCONNECT or @ref GPIO_OUTPUT_PUSH_PULL implementation
+ *          defined. For pins configured as @ref GPIO_OUTPUT_OPEN_DRAIN must
+ *          reflect what is read from the pin, so that a software I2C bus with
+ *          support for clock stretching can be implemented.
+ */
+static inline uword_t gpio_ll_read(gpio_port_t port);
+
+/**
+ * @brief   Perform an `reg |= mask` operation on the I/O register of the port
+ *
+ * @note    The behavior regarding pins not configured as
+ *          @ref GPIO_OUTPUT_PUSH_PULL or as
+ *          @ref GPIO_OUTPUT_OPEN_DRAIN is implementation defined.
+ * @warning Portable code must set the bits in @p mask that do not correspond to
+ *          pins configured as output to zero.
+ *
+ * @param[in] port      port to modify
+ * @param[in] mask      bitmask containing the pins to set
+ */
+static inline void gpio_ll_set(gpio_port_t port, uword_t mask);
+
+/**
+ * @brief   Perform an `reg &= ~mask` operation on the I/O register of the port
+ *
+ * @note    The behavior regarding pins not configured as
+ *          @ref GPIO_OUTPUT_PUSH_PULL or as
+ *          @ref GPIO_OUTPUT_OPEN_DRAIN is implementation defined.
+ * @warning Portable code must set the bits in @p mask that do not correspond to
+ *          pins configured as output to zero.
+ *
+ * @param[in] port      port to modify
+ * @param[in] mask      bitmask containing the pins to clear
+ */
+static inline void gpio_ll_clear(gpio_port_t port, uword_t mask);
+
+/**
+ * @brief   Perform an `reg ^= mask` operation on the I/O register of the port
+ *
+ * @note    The behavior regarding pins not configured as
+ *          @ref GPIO_OUTPUT_PUSH_PULL or as
+ *          @ref GPIO_OUTPUT_OPEN_DRAIN is implementation defined.
+ * @warning Portable code must set the bits in @p mask that do not correspond to
+ *          pins configured as output to zero.
+ *
+ * @param[in] port      port to modify
+ * @param[in] mask      bitmask containing the pins to toggle
+ */
+static inline void gpio_ll_toggle(gpio_port_t port, uword_t mask);
+
+/**
+ * @brief   Helper to use @ref gpio_ll_write side-effect free
+ * @return  Bitmask to or into the value parameter of @ref gpio_ll_write
+ *
+ * See @ref gpio_ll_write on how to use this function
+ */
+static inline uword_t gpio_ll_write_mask(gpio_port_t port);
```

Comment:
This is also new. In most cases the out register of the GPIO peripheral just ignores any writes to pins not configured as output, so this will be implemented as

```C
static inline uword_t gpio_ll_write_mask(gpio_port_t port)
{
    (void)port;
    return 0;
}
```

In that cases, the compiler will optimize out any `data | gpio_ll_write_mask(port)`. But for the ATmega platform this will enable using writing to output pins without messing with the pull resistor config of pins on the same port and without paying a full read-modify-write cycle, if the data to write to the ports can be precalculated.

Reply:
It is a little confusing how the function is used. If I understand the purpose of this function correctly, it should help to avoid side effects that could happen when 1-bits are written to the output register for pins that are not configured as outputs. This would require that the corresponding bits for pins configured as outputs in the mask are set to 1, and all others are set to 0. The value would then have to be ANDed with the mask before writing to the output register to avoid writing 1-bits into the output register for pins that are not configured as outputs.
```
data & gpio_ll_write_mask(port)
```

However, your example code and the test application suggest that 0 is returned for writable bits and you return 0 where I would expect to return `UWORD_MAX` if all bits can be written. IMHO, this would also be valid for ATmega.  If the corresponding bit in the `DDR` is 1 it is an output and the value written to `PORT` defines the level of the pin. Am I wrong?

Reply:
Even more, I would expect side effects on ATmega if `DDR` bits are 0 (configured as input) and 1-bits are written to `PORT`

Reply:
Sadly this wouldn't work. The issue is that writing a 1 bit to an input pin enabled the pull up, writing a zero disables it again. So the state of the output register has to remain as it was before, except for output pins. The idea is that this returns the state of the output register, but clears all bits related to output pins.

The assumption is that the user of the API only changes the output pins anyway (and manages multilexing of output pins on the same port or uses them exclusively - this is also why most users will want to use a high level API instead).

Reply:
> Even more, I would expect side effects on ATmega if DDR bits are 0 (configured as input) and 1-bits are written to PORT

Exactly. But writing zeros to bits corresponding to input pins will also have side effects (disabling the pull ups, in case those were enabled). It really is needed to keep the state of the PORT bits corresponding to input pins at the desired level.

Reply:
Ah OK, you return the bits of `PORT` for pins configured as input in `DDR` to leave them unchanged when writing back to `PORT`. Shouldn't this be better considered in `gpio_ll_write`? For ESP32 I would need such a function with exactly the opposite semantics.

Reply:
Handling it transparently would require two reads (1-2 CPU cycles each) to get the state of PORT and DDR and some computing (`(~DDR & PORT) | (DDR & value)`). That would only yield only a tiny fraction of the speed that is currently achieved - which is square wave at half the CPU clock (1 CPU cycle per write).

Note that this low level API really wants to only expose the hardware in a platform independent way - not to add any convenience to it. The idea is to add the convenience on top of that with a single implementation.

If I understand you correctly on ESP it is illegal to write ones to bits corresponding to input pins. But that is the responsibility of the user to just not do that. A user who does not want to take that responsibility will have to use the convenience pin-based API (the current GPIO API), that will be implemented on top.

Reply:
> Handling it transparently would require two reads (1-2 CPU cycles each) to get the state of PORT and DDR and some computing (`(~DDR & PORT) | (DDR & value)`). That would only yield only a tiny fraction of the speed that is currently achieved - which is square wave at half the CPU clock (1 CPU cycle per write).

This is exactly what I thought why you did it in that way.

What about to define an inline function to generate a valid masked value whatever platform is used instead of getting the write mask and using it correctly? This could be an approach that is platform independent, for example:
```c
static inline uword_t gpio_ll_masked_value(port_t port, uword_t value)
{
    atmega_gpio_port_t *p = (void *)port;
    return (p->port & (~p->ddr)) | value;
}
...
uword_t state1 = gpio_ll_masked_value(port, 1 << pin1);
uword_t state2 = gpio_ll_masked_value(port, 1 << pin2);
```
On another platform this function could be defined
```c
static inline uword_t gpio_ll_masked_value(port_t port, uword_t value)
{
    xyz_gpio_port_t *p = (void *)port;
    return p->oen & value;
}
```
If the function just returns the value, it is doesn't produce additional code
```c
static inline uword_t gpio_ll_masked_value(port_t port, uword_t value)
{
    (void)port
    return value;
}
```

> Note that this low level API really wants to only expose the hardware in a platform independent way - not to add any convenience to it. The idea is to add the convenience on top of that with a single implementation.

:+1:

> If I understand you correctly on ESP it is illegal to write ones to bits corresponding to input pins.

To be honest, it is not really clear to me but my concern is that it might have side effects. Unfortunately, ESP doesn't have such a clear port-based GPIO interface or register mapping as other MCUs. A number of GPIOs are used for internal functions, even though there not broken out. Other GPIOs can be used only as inputs and don't have pull-ups or pull-downs. Other GPIOs are used for essential SoC components like the IROM. In best case, the system crashes when the user tries to write such GPIOs.

> But that is the responsibility of the user to just not do that. A user who does not want to take that responsibility will have to use the convenience pin-based API (the current GPIO API), that will be implemented on top.

 :+1: 

BTW, the ESP32 interface already works as expected, the test application passes, at least without interrupt support so far, but that shouldn't be a problem. The highest square wave frequency is 6.657.789 Hz uncompensated. With `bench` command in `periph_gpio` test, I could about 3.0M times clear/set per second or 2.4M times toggle per  second.

Reply:
I added it, but named it `gpio_ll_prepare_write()` so that it is more clear (at least this is my hope) that this is only needed for the `gpio_ll_write()` function, and not e.g. for `gpio_ll_clear()` or `gpio_ll_set()` (which might do an expensive read-modify-write operation on a register if it is not feasible to implement it in one CPU cycles).